File: 0076-minimum-window-substring/0076-minimum-window-substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str: 
        if len(t) > len(s): return ""      
        freq = [0] * 256
        for ch in t:
            freq[ord(ch)] += 1
        cnt = len(t)
        left, right = 0, 0
        minLen, idx = float("inf"), 0
        while right < len(s):
            if freq[ord(s[right])] > 0:
                cnt -= 1
            freq[ord(s[right])] -= 1            
            while cnt == 0:
                if right - left + 1 < minLen:
                    minLen = right - left + 1
                    idx = left
                freq[ord(s[left])] += 1
                if(freq[ord(s[left])] > 0):
                    cnt += 1
                left += 1
            right += 1
        return s[idx : idx+minLen] if minLen < float('inf') else ""
            
        

File: 0076-minimum-window-substring/0076-minimum-window-substring.java
class Solution {
    public String minWindow(String s, String t) {
        if (s.length() < t.length()) return "";
        HashMap<Character, Integer> freq = new HashMap<>();
        for (int i = 0; i < t.length(); i++) {
            char ch = t.charAt(i);
            freq.put(ch, freq.getOrDefault(ch, 0) + 1);
        }
        int min_l = 0, min_r = 0;
        int cnt = t.length();
        int l = 0, r = 0;
        int shortest = Integer.MAX_VALUE;
        while (r < s.length()) {
            char ch = s.charAt(r);
            if (freq.containsKey(ch)) {
                freq.put(ch, freq.get(ch) - 1);
                if (freq.get(ch) >= 0) cnt--;
            }
            while (cnt == 0) {                
                if (r - l + 1 < shortest) {
                    shortest = r - l + 1;
                    min_l = l;
                    min_r = r;
                }
                if (freq.containsKey(s.charAt(l))) {
                    freq.put(s.charAt(l), freq.get(s.charAt(l)) + 1);
                    if (freq.get(s.charAt(l)) > 0) cnt++;                    
                }
                l++;
            }
            r++;
        }
        
        if (shortest == Integer.MAX_VALUE) return "";
        return s.substring(min_l, min_r + 1);
    }
}